File: leetCode/0015_3sum.js
// 15. 3Sum

var threeSum = function(nums) {
  nums = nums.sort((a, b) => a-b);
  let ret = [];
  for ( let i = 0 ; i < nums.length-2 ; i++ ) {
    if ( nums[i] == nums[i-1] ) continue;
    let l = i+1;
    let r = nums.length-1;
    let a = nums[i];

    while( l < r ) {
      if ( a + nums[l] + nums[r] == 0 ) {
        ret.push([a, nums[l], nums[r]]);
        while(nums[++l] == nums[l-1]){};
        while(nums[--r] == nums[r+1]){};
      } else if ( a + nums[l] + nums[r] < 0 ) {
        while(nums[++l] == nums[l-1]){};
      } else {
        while(nums[--r] == nums[r+1]){};
      }
    }
  }
  return ret;
};

File: leetCode/0015_3Sum.java
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);

        List<List<Integer>> ret = new ArrayList<List<Integer>>();

        for ( int i = 0 ; i <= nums.length - 3 ; i++ ) {
            if ( i > 0 && nums[i] == nums[i-1] ) continue;

            int l = i+1;
            int r = nums.length - 1;
            int target = 0 - nums[i];

            while ( l < r ) {
                if ( nums[l] + nums[r] == target ) {
                    ret.add(Arrays.asList(nums[i], nums[l], nums[r]));
                    while ( l < r && nums[++l] == nums[l-1] ) {};
                    while ( l < r && nums[--r] == nums[r+1] ) {};
                } else if ( nums[l] + nums[r] < target ) {
                    l++;
                } else {
                    r--;
                }
            }
        }
        
        return ret;
    }
}